cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp

```diff
@@ -0,0 +1,719 @@
+		template<class... Args> iterator emplace_hint(const_iterator position, Args&&... args);
+		pair<iterator, bool> insert(const value_type& obj);
+		pair<iterator, bool> insert(value_type&& obj);
+		iterator insert(const_iterator hint, const value_type& obj);
+		iterator insert(const_iterator hint, value_type&& obj);
+		template<class InputIterator> void insert(InputIterator first, InputIterator last); 
+
+		iterator erase(iterator position);
+		iterator erase(const_iterator position);
+		iterator erase(const_iterator first, const_iterator last);
+		void swap(unordered_set&) noexcept/*(allocator_traits<Allocator>::is_always_equal::value && is_nothrow_swappable_v<Hash> && is_nothrow_swappable_v<Pred>)*/;
+		void clear() noexcept;
+
+		template<class H2, class P2> void merge(unordered_set<Key, H2, P2, Allocator>& source);
+		template<class H2, class P2> void merge(unordered_set<Key, H2, P2, Allocator>&& source);
+
+		iterator find(const key_type& k);
+		const_iterator find(const key_type& k) const;
+		pair<iterator, iterator> equal_range(const key_type& k);
+		pair<const_iterator, const_iterator> equal_range(const key_type& k) const;
+	};
+}
+
+std::vector<std::vector<int>> returnValue();
+std::vector<std::vector<int>>& returnRef();
+
+std::vector<std::vector<int>> external_by_value(std::vector<std::vector<int>>);
+
+std::vector<std::vector<int>> external_by_const_ref(const std::vector<std::vector<int>>&);
+
+// *: Will be detected once extract destruction of unnamed temporaries and generate IR for them
+
+const std::vector<std::vector<int>>& return_self_by_ref(const std::vector<std::vector<int>>& v) {
+	return v;
+}
+
+std::vector<std::vector<int>> return_self_by_value(const std::vector<std::vector<int>>& v) {
+	return v;
+}
+
+void test() {
+  for (auto x : returnValue()) {} // GOOD
+  for (auto x : returnValue()[0]) {} // BAD [NOT DETECTED] (see *)
+  for (auto x : external_by_value(returnValue())) {} // GOOD
+  for (auto x : external_by_const_ref(returnValue())) {} // GOOD
+  for (auto x : returnValue().at(0)) {} // BAD [NOT DETECTED] (see *)
+
+  for (auto x : returnRef()) {} // GOOD
+  for (auto x : returnRef()[0]) {} // GOOD
+  for (auto x : returnRef().at(0)) {} // GOOD
+
+	for(auto it = returnValue().begin(); it != returnValue().end(); ++it) {} // BAD
+
+	{
+	auto v = returnValue();
+	for(auto it = v.begin(); it != v.end(); ++it) {} // GOOD
+	}
+
+	{
+	auto&& v = returnValue();
+	for(auto it = v.begin(); it != v.end(); ++it) {} // GOOD
+	}
+
+	{
+	auto&& v = returnValue()[0];
+	for(auto it = v.begin(); it != v.end(); ++it) {} // BAD [NOT DETECTED] (see *)
+	}
+	
+	{
+	auto&& v = returnRef();
+	for(auto it = v.begin(); it != v.end(); ++it) {} // GOOD
+	}
+
+	{
+	auto&& v = returnRef()[0];
+	for(auto it = v.begin(); it != v.end(); ++it) {} // GOOD
+	}
+
+	for (auto x : return_self_by_ref(returnValue())) {} // BAD [NOT DETECTED] (see *)
+
+	for (auto x : return_self_by_value(returnValue())) {} // GOOD
```

Comment:
How confident are you that each of these test cases are annotated correctly?  Would it be worth one of us trying to prove actual failures for each case?

Reply:
I tested them all on https://ideone.com/ before pushing these annotations 😄

Reply:
And the annotations also matches my understanding of the semantics of temporary objects, at least. So that's also a bonus!

Reply:
Awesome, I appreciate the thoroughness!